include/zephyr/arch/riscv/csr.h

```diff
@@ -233,4 +234,59 @@
 				: "memory");			\
 })
 
+/* uhartid -- user hardware thread id, seems to be not yet defined in toolchain */
+#ifdef CONFIG_RISCV_KERNEL_IN_USER_MODE
+	/* zephyr runs in user mode */
+
+	/* inline assembly defines */
+	#define XSTATUS_IEN			USTATUS_IEN
+	#define XSTATUS_DEF_RESTORE		USTATUS_DEF_RESTORE
+	#define XSTATUS				0x000
+
+	/* register definition for assembly */
+	#define	uhartid				0x014
+
+	#define xhartid				uhartid
+	#define xscratch			uscratch
+	#define xstatus				ustatus
+	#define xepc				uepc
+	#define xcause				ucause
+	#define xret				uret
+
+#elif CONFIG_RISCV_KERNEL_IN_SUPERVISOR_MODE
+	/* zephyr runs in supervisor mode */
+
+	/* inline assembly defines */
+	#define XSTATUS_IEN			SSTATUS_IEN
+	#define XSTATUS_DEF_RESTORE		SSTATUS_DEF_RESTORE
+	#define XSTATUS				0x000
+
+	/* register definition for assembly */
+	#define	shartid				0x7D2
+
+	#define xhartid				shartid
```

Comment:
I'm wondering which RV spec version it is based. Cause current RV specs don't define shartid.

What we can do is to store CPU ID to per-core cpu structure during boot. We cache the pointer to current cpu structure in `xscratch` register - so we can easily get CPU ID latter.

Reply:
Hey @evgeniy-paltsev 
> I'm wondering which RV spec version it is based. Cause current RV specs don't define shartid.

yep, shartid/ uhartid are non standard and shouldnt be used. 

> 
> What we can do is to store CPU ID to per-core cpu structure during boot. We cache the pointer to current cpu structure in `xscratch` register - so we can easily get CPU ID latter.

I am in the middle of doing this right now :)

Reply:
just wondering if openSBI will also start multi core cluster, ... or zephyr is responsible to do this.

Reply:
> just wondering if openSBI will also start multi core cluster, ... or zephyr is responsible to do this.

In SMP, Zephyr's boot hart is responsible for bringing up the other harts. So when OpenSBI hands control over to Zephyr, we at least have the boot hartid in the a0 register. That's as far as I am at the moment too. 

I think by design, on RISC-V (privileged spec) there is no concept of hartid outside of machine mode. As in, you shouldn't need to know what the hartid is. As such, I don't think we can include aliases for shartid and uhartid, even if they are available on your platform.

I think we can achieve what we need by storing the boot hartid in sscratch (supervisor mode). I think we need to make a couple of assumptions in this respect: the first is, if using SBI, the platform's PMP are configured and the SBI payload knows which cores that belong to Zephyr. This at least allows Zephyr's boot code to function. The challenge will be the PLIC and timer driver.

Reply:
agree to some extend. by now it would be ok to start only 1 core. later we should think about starting all cores. for me it looks like that openSBI has to hand over some kind of structure, which indicates which cores to use and boot.

PLIC is something that can be handled by openSBI only -- I guess. because machine mode is responsible to forward to supervisor mode. in supervisor mode we only have the interrupt vector tables and the supervisor global enable and peripheral enable irqs.

regarding timer, the machine timer is no option anymore, this is definetly handled by openSBI. I have a suggested implementation in the PR. actually it should be possible to use exakt same driver for your board.

Reply:
btw: @con-pax did you check out my qemu example. I m pretty happy to have it running, ... https://github.com/zephyrproject-rtos/zephyr/issues/68133#issuecomment-2304523249

Reply:
> btw: @con-pax did you check out my qemu example. I m pretty happy to have it running, ... [#68133 (comment)](https://github.com/zephyrproject-rtos/zephyr/issues/68133#issuecomment-2304523249)

I'm going to get to this once I finish my awful rebase 😄

Reply:
> PLIC is something that can be handled by openSBI only -- I guess. because machine mode is responsible to forward to supervisor mode. in supervisor mode we only have the interrupt vector tables and the supervisor global enable and peripheral enable irqs. thus I disabled PLIC in qemu. using only the interrupt vectors.

Thing is, Linux implements a PLIC driver and runs in s-mode, I believe it can be done